atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventLog.java

```diff
@@ -51,31 +52,36 @@ CacheUpdate processUpdate(LockWatchStateUpdate update) {
     }
 
     /**
-     * @param lastKnownVersion latest version that the client knows about; should be before timestamps in the mapping;
-     * @param endVersion       mapping from timestamp to identified version from client-side event cache;
+     * @param versionBounds contains:
+     *                     - the latest version the client knows about, which should be before the timestamps in the
+     *                     mapping;
+     *                     - the end version, which is the upper bound of events that should be returned;
+     *                     - the earliest version to which the events may be condensed to in the case of a snapshot.
      * @return lock watch events that occurred from (exclusive) the provided version, up to the end version (inclusive).
+     * This may begin with a snapshot if the latest version is too far behind.
      */
-    public ClientLogEvents getEventsBetweenVersions(
-            Optional<LockWatchVersion> lastKnownVersion,
-            LockWatchVersion endVersion) {
-        Optional<LockWatchVersion> startVersion = lastKnownVersion.map(this::createStartVersion);
-        LockWatchVersion currentVersion = getLatestVersionAndVerify(endVersion);
+    public ClientLogEvents getEventsBetweenVersions(VersionBounds versionBounds) {
+        Optional<LockWatchVersion> startVersion = versionBounds.startVersion().map(this::createStartVersion);
+        LockWatchVersion currentVersion = getLatestVersionAndVerify(versionBounds.endVersion());
 
         if (!startVersion.isPresent() || differentLeaderOrTooFarBehind(currentVersion, startVersion.get())) {
+            long snapshotVersion = versionBounds.snapshotVersion();
+            Collection<LockWatchEvent> afterSnapshotEvents = eventStore.getEventsBetweenVersionsInclusive(
```

Comment:
Let's assume that the user is on version `U`, the cache's earliest is `V` (so its internal snapshot is on that version), and the timestamp mapping contains versions from `W` to `X`.

Case 1: `snapshotVersion < endVersion` (i.e. `W < X`):
The condensed snapshot will contain the information from `V`, the base snapshot, up to and including `W` - therefore transactions on this earliest version will contain exactly the info they need, and any on >`W` will have the additional events that they need (which is computed here, events `W + 1` -> `X`).

Case 2: `snapshotVersion == endVersion` (or `W == X`):
In this case, there are no further events we need to retrieve. The extra change in the `eventStore` makes asking for `X+1, X` a valid request (where before it would throw); I felt it just made sense to push that down so we don't have to worry about it here.

That should be all cases covered.